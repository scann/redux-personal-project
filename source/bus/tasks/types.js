export const types = {
    //Sync
    FILL_TASKS: 'FILL_TASKS',

    //Async
    FETCH_TASKS_ASYNC: 'FETCH_TASKS_ASYNC',
};
