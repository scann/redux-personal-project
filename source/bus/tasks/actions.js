//Types
import { types } from './types';

export const tasksActions = {
    //Sync
    fillTasks: (tasks) => {
        return {
            type:    types.FILL_TASKS,
            payload: tasks,
        };
    },
    createTask: (message) => {
        return {
            type:    types.CREATE_TASK,
            payload: message,
        };
    },

    //Async
    fetchTasksAsync: () => {
        return {
            type: types.FETCH_TASKS_ASYNC,
        };
    },
    createTaskAsync: (message) => {
        return {
            type:    types.CREATE_TASK_ASYNC,
            payload: message,
        };
    },
};
