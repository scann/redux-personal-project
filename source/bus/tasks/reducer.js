//Core
import { fromJS, List } from 'immutable';

//Instruments
import { types } from './types';

const initialState = List();

export const tasksReducer = (state = initialState, action) => {
    switch (action.type) {
        case types.FILL_TASKS:
            return fromJS(action.payload);

        case types.CREATE_TASK:
            return state.unshift(fromJS(action.payload));

        case types.REMOVE_TASK:
            const taskId = fromJS(action.payload);

            return state.filter((task) => task.get('id') !== taskId);

        case types.COMPLETE_ALL_TASKS:

            return state.map((task) => task.set('completed', true));

        case types.UPDATE_TASK:
            return state.update(
                state.findIndex(
                    (task) => task.get('id') === action.payload.id),
                () => fromJS(action.payload)
            );

        default:
            return state;
    }
};
