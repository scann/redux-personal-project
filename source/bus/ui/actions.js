//Types
import { types } from './types';

export const uiActions = {
    startSpinning: () => {
        return {
            type: types.START_SPINNING,
        };
    },
    stopSpinning: () => {
        return {
            type: types.STOP_SPINNING,
        };
    },
    emitError: (error, meta = null) => {
        return {
            type:    types.EMIT_ERROR,
            payload: error,
            error:   true,
            meta,
        };
    },
    updateTasksFilter: (filter) => {
        return {
            type:    types.UPDATE_TASKS_FILTER,
            payload: filter,
        };
    },
    startEditingTask: (id, message) => {
        return {
            type:    types.START_EDITING_TASK,
            payload: {
                id,
                message,
            },
        };
    },
    resetEditingTask: () => {
        return {
            type: types.RESET_EDITING_TASK,
        };
    },
    confirmEditingTask: (taskId, message) => {
        return {
            type:    types.CONFIRM_EDITING_TASK,
            payload: {
                taskId,
                message,
            },
        };
    },
    updateTaskMessage: (message) => {
        return {
            type:    types.UPDATE_TASK_MESSAGE,
            payload: message,
        };
    },
};
