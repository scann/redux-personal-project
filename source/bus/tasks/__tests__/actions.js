import { tasksActions } from '../../tasks/actions';

describe('tasks actions:', () => {
    test('fillTasks', () => {
        expect(tasksActions.fillTasks(__.tasksList)).toMatchSnapshot();
    });

    test('createTask', () => {
        expect(tasksActions.createTask(__.testMessage.message)).toMatchSnapshot();
    });

    test('removeTask', () => {
        expect(tasksActions.removeTask(__.testMessage.id)).toMatchSnapshot();
    });

    test('completeAllTasks', () => {
        expect(tasksActions.completeAllTasks()).toMatchSnapshot();
    });

    test('updateTask', () => {
        expect(tasksActions.updateTask(__.testMessage)).toMatchSnapshot();
    });

    test('fetchTasksAsync', () => {
        expect(tasksActions.fetchTasksAsync()).toMatchSnapshot();
    });

    test('createTaskAsync', () => {
        expect(tasksActions.createTaskAsync(__.testMessage.message)).toMatchSnapshot();
    });

    test('removeTaskAsync', () => {
        expect(tasksActions.removeTaskAsync(__.testMessage.id)).toMatchSnapshot();
    });

    test('completeAllTasksAsync', () => {
        expect(tasksActions.completeAllTasksAsync()).toMatchSnapshot();
    });

    test('updateTaskAsync', () => {
        expect(tasksActions.updateTaskAsync(__.testMessage)).toMatchSnapshot();
    });
});
