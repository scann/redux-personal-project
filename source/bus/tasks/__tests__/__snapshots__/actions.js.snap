// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tasks actions: completeAllTasks 1`] = `
Object {
  "type": "COMPLETE_ALL_TASKS",
}
`;

exports[`tasks actions: completeAllTasksAsync 1`] = `
Object {
  "type": "COMPLETE_ALL_TASKS_ASYNC",
}
`;

exports[`tasks actions: createTask 1`] = `
Object {
  "payload": "TEST_MESSAGE",
  "type": "CREATE_TASK",
}
`;

exports[`tasks actions: createTaskAsync 1`] = `
Object {
  "payload": "TEST_MESSAGE",
  "type": "CREATE_TASK_ASYNC",
}
`;

exports[`tasks actions: fetchTasksAsync 1`] = `
Object {
  "type": "FETCH_TASKS_ASYNC",
}
`;

exports[`tasks actions: fillTasks 1`] = `
Object {
  "payload": Object {
    "tasks": Array [
      Object {
        "completed": true,
        "favorite": true,
        "id": "TEST_ID1",
        "message": "TEST_MESSAGE1",
      },
      Object {
        "completed": false,
        "favorite": false,
        "id": "TEST_ID2",
        "message": "TEST_MESSAGE2",
      },
    ],
  },
  "type": "FILL_TASKS",
}
`;

exports[`tasks actions: removeTask 1`] = `
Object {
  "payload": "TEST_ID",
  "type": "REMOVE_TASK",
}
`;

exports[`tasks actions: removeTaskAsync 1`] = `
Object {
  "payload": "TEST_ID",
  "type": "REMOVE_TASK_ASYNC",
}
`;

exports[`tasks actions: updateTask 1`] = `
Object {
  "payload": Object {
    "id": "TEST_ID",
    "message": "TEST_MESSAGE",
  },
  "type": "UPDATE_TASK",
}
`;

exports[`tasks actions: updateTaskAsync 1`] = `
Object {
  "payload": Object {
    "id": "TEST_ID",
    "message": "TEST_MESSAGE",
  },
  "type": "UPDATE_TASK_ASYNC",
}
`;
